# COMP 641 Group Programming Project    

**Project Title:** Predicting 30-Day Readmission in Diabetes Patients    

__________________________________________________________________________________________________________________
**Group Members:**  
Aaron Hofman  
Arunachalesh M Kembhavimath  
Phone Pyae Zaw  
Jaztin Tabunda  

__________________________________________________________________________________________________________________

## Data Science Project Instructions – Master’s Program

Course Title: COMP 641 – Fundamentals of Data Science
Project Title:  End to End Data Science Project
Due Date: December 12, 2025

Overview
This project is a capstone-style assignment designed to apply the core concepts of data science covered throughout the course. You will identify a real-world problem, acquire and prepare data, conduct exploratory and statistical analysis, build and evaluate models, and present your findings in both technical and non-technical formats.

Objectives
By completing this project you will:
        Frame a real-world problem in data science terms.
        Collect, clean, and analyze data.
        Apply data science methods (EDA, modeling, evaluation).
        Communicate results effectively to stakeholders.
        Practice reproducibility and ethical data practices.

Proposal  Submission: (Due: Week 3)

Submit a one-page proposal including:
1. Project title and team members
2. Problem statement
3. Source of data (must be accessible)
4. Planned methods/techniques (e.g., regression, classification, clustering)
5. Ethical considerations

Project Components

Data Collection & Cleaning

    Students must acquire data from a reliable source, such as public datasets, APIs, or through web scraping techniques.
    Any challenges encountered during the data sourcing process should be clearly documented and explained.
    Students must clean and preprocess the data, including handling missing values and performing appropriate feature engineering steps.

Exploratory Data Analysis (EDA)

    Students are required to visualize data trends, identify outliers, and explore data distributions to gain insight into the dataset.
    At least three visualizations must be included, using tools such as matplotlib, seaborn, or similar libraries.
    A written narrative must accompany the visualizations, summarizing key findings from the exploratory analysis.

Modeling & Analysis

    Students must choose appropriate machine learning models based on the nature of their problem, such as regression, clustering, or classification.
    The dataset should be split into training and testing sets, or evaluated using cross-validation techniques.
    Model performance must be assessed using relevant evaluation metrics, such as accuracy, RMSE, or AUC.

Ethics & Bias Discussion

    Students must identify and discuss any potential biases present in the data collection process or modeling approach.
    The discussion should include the possible impact of these biases on stakeholders or end users of the system.

Presentation (Due: December 12)

Students are required to deliver a 5–10 minute in-class presentation summarizing their project. The presentation should clearly explain the problem addressed, the data used, the methods applied, and what are the key findings. Visual aids such as charts, graphs, or dashboards must be included to support the narrative. The presentation should be accessible to a broad audience and avoid excessive technical jargon.

 

Final Report (Due: December 12)

            Students must submit a technical report that is between 5 to 10 pages in length.
            The report should include an introduction and clearly framed problem statement. A detailed description of the data and methodology used must be provided in the report. Students must present and interpret the results of their analysis and modeling.


## References/Sources

Dataset Files  
https://archive.ics.uci.edu/dataset/296/diabetes+130-us+hospitals+for+years+1999-2008  

ICD9 Codes  
https://www.aapc.com/codes/icd9-codes-range/  
https://www.aapc.com/codes/icd9-codes-range/136/  
https://www.aapc.com/codes/icd9-codes-range/152/  

## Project Environment Preparation

### Project Imports:

In [8]:
# from ucimlrepo import fetch_ucirepo
import pandas as pd
import os
from tqdm.auto import tqdm
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

### Preparing CSV File paths for loading and Output Purposes:

This is just to set up the paths where we will be saving the encoded dataset for the model and the place where the csv file is

In [11]:
# Location of diabetic_data.csv.
# "<insert file path of csv file/dataset>"
import_df_location = "C:\\Users\\Jaztin T\\Downloads\\COMP-641-ProjectJupyterNotebook\\diabetic_data.csv"
# Locations for output files.
# "<insert file path of where you want output files to be>"
file_save_loaction = "C:\\Users\\Jaztin T\\Downloads\\COMP-641-ProjectJupyterNotebook"

# File location provided check.
if (import_df_location == ""):
    raise ValueError("import_df_location cannot be empty.")
if (file_save_loaction == ""):
    raise ValueError("file_save_loaction cannot be empty.")

### Setting up Variables for Output Text Information Files:

The following variables are used to produce the Output text files at the end of the notebook file

In [14]:
# Output text.
list_of_unique_and_nunique = ""
check_duplicates_in_first_two = ""
replacing_race_with_integers = ""
replacing_gender_with_integers = ""
replace_age_categories_with_integers = ""
removing_weight_column = ""
condensed_insurance_information = ""
replacing_medical_specialty_with_integers = ""
replacing_diagnoses_codes_with_integers = ""
replacing_medication_used_indication_with_integers = ""
replacing_max_glu_serum_with_integers = ""
replacing_A1Cresult_with_integers = ""
replacing_change_with_integers = ""
replacing_diabetesMed_with_integers = ""
replacing_readmitted_with_integers = ""

## Data Collection & Cleaning

### Data Loading 

The Main Diabetes Dataset is from diabetic_data.csv. This includes the information on hospital readmissions and patients.    

The other datasets are originally from the IDS_mapping.csv  file. These datasets have the meaning of some columns values in the Main Diabetes Dataset. They were split from the IDS_mapping.csv to their own .csv files. Then they were loaded in to their respective dataframes for ease of use later in the EDA stage.

In [18]:
# Importing datasets and loading them into their respective dataframes (Assuming it's in the same directory as jupyter notebook file)
# Main Diabetes dataset
DiabetesDF = pd.read_csv('diabetic_data.csv')

# Admission Source Type dataset. Derived from IDS_mapping.csv
# combined with the main diabetes dataset, one can tell the origin of the patient prior to hospital admission
admissionSourceDF = pd.read_csv('admission_source_types.csv')

# Admission Type dataset. Derived from IDS_mapping.csv
# combined with the main diabetes dataset, one can tell the patient's type of visit to the hospital (e.g. urgent, emergency, elective)
admissionTypeDF = pd.read_csv('admission_type.csv')

# Discharge disposition type dataset. Derived from IDS_mapping.csv
# combined with the main diabetes dataset, one can tell where the patient went after discharge from hospital
dischargeDispositionDF = pd.read_csv('discharge_disposition.csv')

# dictionary of dataframes in case we want to loop through each dataframe later
dataFrames = {
    "Diabetes Dataset": DiabetesDF,
    "Admission Source Dataset": admissionSourceDF,
    "Admission Type Dataset": admissionTypeDF,
    "Discharge Disposition Dataset": dischargeDispositionDF
}

### Data Initial Inspection - Seeing what we are dealing with

#### Summary/Overview of Each Dataframe (Prior to Cleaning)

In [21]:
# for each dataframe we have, we'll get a brief summary
for name, dataFrame in dataFrames.items():
    print(name + "----------------------------------------------\n")
    print(dataFrame.info())
    print("----------------------------------------------\n")

Diabetes Dataset----------------------------------------------

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      101766 non-null  object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    101766 non-null  object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                101766 non-null  object
 11  medical_specialty         101766 non-null  object
 12  num_lab_procedures        101766 non-null  int64 


Looking at each dataset, we find information regarding each dataset's features, data types, number of rows/entries, number of columns, and whether or not the datasets contain 'null' values. Firstly, one can see that the shape (rows x columns) of the diabetes dataset is 101766 rows/entries by 50 columns/features. The dataset contains demographics, encounters, and lab results, with values stored as integers and objects/strings. Besides this, at a glance, one may notice that most columns in the diabetes dataset have "non-null" values in them. However, there are still columns in the dataset that contain "null" values. Namely, **max_glu_serum** and **A1Cresult** clearly have null values in their columns. To go into specifics, **max_glu_serum** has 5346 non-null values and **A1Cresult** has 17018 non-null values out of the 101766 rows/entries. Besides the main diabetes dataset, it's important to note that we have 1 'null' value in every other dataset's "description" column. In other words, once we merge the mapping datasets to the main diabetes dataset, one would most likely find more null values

#### Descriptive Statistics of the Diabetes Dataset (Prior to Cleaning)

In [24]:
# we'll look at the numerical and categorical stats for the Diabetes DF 
print("Diabetes DF's numerical stats ----------------------------------------------\n")
print(DiabetesDF.describe())
print("\nDiabetes DF's categorical stats ----------------------------------------------\n")
# ensure we show the descriptive statistics for all columns
pd.set_option('display.max_columns', None)
print(DiabetesDF.describe(include="object"))
print("----------------------------------------------\n")

Diabetes DF's numerical stats ----------------------------------------------

       encounter_id   patient_nbr  admission_type_id  \
count  1.017660e+05  1.017660e+05      101766.000000   
mean   1.652016e+08  5.433040e+07           2.024006   
std    1.026403e+08  3.869636e+07           1.445403   
min    1.252200e+04  1.350000e+02           1.000000   
25%    8.496119e+07  2.341322e+07           1.000000   
50%    1.523890e+08  4.550514e+07           1.000000   
75%    2.302709e+08  8.754595e+07           3.000000   
max    4.438672e+08  1.895026e+08           8.000000   

       discharge_disposition_id  admission_source_id  time_in_hospital  \
count             101766.000000        101766.000000     101766.000000   
mean                   3.715642             5.754437          4.395987   
std                    5.280166             4.064081          2.985108   
min                    1.000000             1.000000          1.000000   
25%                    1.000000             1.0

Looking at the descriptive statistics for the numerical and categorical data, we can list some of the visible observations and problems with the dataset.    

Observations:
- There are 8 categories for admission type
- There are 28 categories for discharge disposition types
- There are 25 categories for admission source type
- The minimum number of days a patient is in the hospital is 1 day. The maximum number of days is 14 days. The average is about 4 (mean = 4.395).
- The min number of lab procedures in one visit is 1. The max is 132 lab procedures in one visit. The average is about 43 (mean = 43.09).
- The min number of procedures besides lab procedures in one visit is 0. The max is 6. The average is about 1 (mean = 1.339).
- The min number of medications administered at the time of encounter is 0. The max is 81. The average is about 16 (mean = 16.021). 
- The min number of outpatient visits of the patient in the year preceding the encounter is 0. The max is 42. The average is about 0 (mean = 0.369). 
- The min number of emergency visits of the patient in the year preceding the encounter is 0. The max is 76. The average is about 0 (mean = 0.197).
- The min number of inpatient visits of the patient in the year preceding the encounter is 0. The max is 21. The average is about 0 (mean = 0.635).
- The min number of diagnoses at the time of encounter is 1. The max is 16. The average is about 7 (mean = 7.422).
- There are 6 different races categories
- There are 3 different gender categories
- There are 10 different age categories
- There are 10 different weight categories
- There are 18 different payer code categories
- There are 73 different medical specialty categories
- There are more than 700 different diagnostic categories for diag 1, 2 and 3
- There are 3 different categories for lab results (A1CResult, max_glu_serum)
- There is a maximum of 4 categories for drug-prescribed related columns (metformin, repaglinide, etc.)
- There are 2 different categories regarding change in diabetes prescription (change, and diabetesMed) 
- There are 3 categories for the readmitted category

Problems:
- Significant amount of Null values in A1CResult
- Significant amount of Null values in max_glu_serum
- "?" values in some columns (weight, payer_code, medical_specialty, etc.) are the most frequent occurring values. In other words, these would need to be changed to null values to properly calculate the number of null elements in the dataset. 
- Readmitted in 30 days column has 3 categories. Must convert to 2 categories (Readmitted within 30 days or not). Might involve dropping the rows or consolidating 2 categories as one.
- Examide and Citoglipton columns may be dropped since the entire dataset is one category for those columns


#### First Five Elements of each Dataset (Prior to Cleaning)

In [27]:
for name, dataFrame in dataFrames.items():
    print(name + "----------------------------------------------\n")
    print(f"{dataFrame.head(5)} \n")
    print("----------------------------------------------\n")

Diabetes Dataset----------------------------------------------

   encounter_id  patient_nbr             race  gender      age weight  \
0       2278392      8222157        Caucasian  Female   [0-10)      ?   
1        149190     55629189        Caucasian  Female  [10-20)      ?   
2         64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3        500364     82442376        Caucasian    Male  [30-40)      ?   
4         16680     42519267        Caucasian    Male  [40-50)      ?   

   admission_type_id  discharge_disposition_id  admission_source_id  \
0                  6                        25                    1   
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   

   time_in_hospital payer_code         medical_specialty  num_lab_proc

Upon looking at the first five elements of each data, one can see the data we are dealing with in each column, and some obvious issues, as previously seen. One obvious issue is the fact that the first 5 rows lack a weight value in the Diabetes dataset. In other words, the weight column should most likely be excluded since it doesn't look like we can get valuable information from it. Additionally, it shows that one would have to look out for "?" values in the dataset. Besides the issues we've seen already, it looks like we'll have to look into translating the diagnostic columns/features or dropping the the columns entirely because we aren't given any other information/dataset other than that it's an ICD9 code.

#### Missing value counts per column (Prior to Cleaning)

In [30]:
for name, dataFrame in dataFrames.items():
    print(name + "'s Missing Value Counts Per Column:-------------------------------------------------------------\n")
    print(str(dataFrame.isna().sum()) + "\n")

Diabetes Dataset's Missing Value Counts Per Column:-------------------------------------------------------------

encounter_id                    0
patient_nbr                     0
race                            0
gender                          0
age                             0
weight                          0
admission_type_id               0
discharge_disposition_id        0
admission_source_id             0
time_in_hospital                0
payer_code                      0
medical_specialty               0
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
number_emergency                0
number_inpatient                0
diag_1                          0
diag_2                          0
diag_3                          0
number_diagnoses                0
max_glu_serum               96420
A1Cresult                   84748
metformin                       0
repaglinide                     0
na

Here we have the "missing value counts per column" which will look for the number of null values in the dataset. From just glancing at the "missing value counts per column", one might come to the conclusion that only max_glu_serum (94% null) and A1Cresult (83% null) have missing values. However, as stated previously, we need to also consider that other columns in the diabetes dataset mark empty data cells with "?" mark. With that being said, merging the other datasets into the main diabetes datasets will be done to get a more accurate number of "null" values. Additionally, the count will include looking at rows with "?"

#### Looking at the options for each Category in all datasets (Prior to Cleaning)

In [33]:
for name, dataFrame in dataFrames.items():
    for column in dataFrame.select_dtypes(include=['object']):
        print(f"{column}'s categories from {name} dataset:")
        print(dataFrame[column].unique())
        print("\n")

race's categories from Diabetes Dataset dataset:
['Caucasian' 'AfricanAmerican' '?' 'Other' 'Asian' 'Hispanic']


gender's categories from Diabetes Dataset dataset:
['Female' 'Male' 'Unknown/Invalid']


age's categories from Diabetes Dataset dataset:
['[0-10)' '[10-20)' '[20-30)' '[30-40)' '[40-50)' '[50-60)' '[60-70)'
 '[70-80)' '[80-90)' '[90-100)']


weight's categories from Diabetes Dataset dataset:
['?' '[75-100)' '[50-75)' '[0-25)' '[100-125)' '[25-50)' '[125-150)'
 '[175-200)' '[150-175)' '>200']


payer_code's categories from Diabetes Dataset dataset:
['?' 'MC' 'MD' 'HM' 'UN' 'BC' 'SP' 'CP' 'SI' 'DM' 'CM' 'CH' 'PO' 'WC' 'OT'
 'OG' 'MP' 'FR']


medical_specialty's categories from Diabetes Dataset dataset:
['Pediatrics-Endocrinology' '?' 'InternalMedicine'
 'Family/GeneralPractice' 'Cardiology' 'Surgery-General' 'Orthopedics'
 'Gastroenterology' 'Surgery-Cardiovascular/Thoracic' 'Nephrology'
 'Orthopedics-Reconstructive' 'Psychiatry' 'Emergency/Trauma'
 'Pulmonology' 'Surgery-Neu

Looking at the options in each category, one may find more issues with the dataset.

Problems:
- **Gender** category contains 'Unknown/Invalid' category which counts as a "null" value. Most likely will change to NaN
- **Description** categories from the IDF datasets have categories that are synonymous to null ('Not Mapped', 'Not Available', 'NULL'). Most likely will change to NaN
- We see the "?" category in multiple columns again. Most likely will change to NaN
- **Discharge Disposition** contains a value that isn't a discharge. "Admitted as an inpatient to this hospital" will be handled when the categories are consolidated.
- The **medical_specialty**, and **diagnostic** (diag_1, diag_2, diag_3) categories have too many options. Most likely needs to be consolidated.
- The **readmitted** catgeories is 3 categories:  
NO - No record of readmission, >30 - Greater than 30 days readmission, <30 - Less than 30 days readmission.  
Most likely will need to make this 2 categories for the machine learning model:  
Yes - readmitted in 30 days, No - Not readmitted in 30 days. However, we can just leave it right now for EDA Visualization.  

##### For Output Text Information File

In [36]:
# Method that prints both a list of the unique elements and the number of unique elements from a provided column name.
# The user can also state when this print method was used.
def unique_element_and_count_string(column_name, before_or_after_or_empty):
    string_to_return = f"Unique Elements {before_or_after_or_empty}: {str(DiabetesDF[column_name].unique())}\n"
    string_to_return += f"Unique Element Count {before_or_after_or_empty}: {str(DiabetesDF[column_name].nunique())}\n"
    return string_to_return

column_names_list = DiabetesDF.columns.tolist()
# Save list and count of unique elements from each column.
list_of_unique_and_nunique += "List and count of unique elements from each column in dataframe.\n"
for element in column_names_list:
    list_of_unique_and_nunique += f"Column Name: {element}\n"
    list_of_unique_and_nunique += unique_element_and_count_string(element, "")
    list_of_unique_and_nunique += "\n"

#### Checking for Duplicate Entries (Prior to Cleaning)

In [38]:
# each entry in the dataset has a unique patient id and encounter id
# a duplicate entry in this case would be if there are multiple rows with the same patient id AND encounter id

# we use the duplicated method to check for any duplicates entries that have the same patient id AND encounter id. 
series = DiabetesDF.duplicated(subset=['encounter_id', 'patient_nbr'])

# any() will return true if a duplicate is found
print(series.any())


False


A duplicate entry would mean that a patient id and encounter id would be the same as a different row. A patient can come back multiple times to the same hospitals, so checking by just the patient id is not sufficient to find duplicates. In this case, there are no duplicates in the dataset.

##### For Output Text Information File

In [41]:
# Checking for duplicates in encounter_id and patient_nbr.
first_two_column_names = column_names_list[:2]
list_of_groups = []
for element in first_two_column_names:
    current_duplicate_check = DiabetesDF.groupby(element)
    list_of_groups.append(current_duplicate_check)
check_duplicates_in_first_two += "During earlier analysis, duplicate elements were found in patient_nbr.\n\n"

### Making Adjustments/Cleaning the Dataframes 

From the initial inspection of the dataset, we identified the following problems:  
- Missing/Null Values  
  - A1Cresult, weight, max_glu_serum have a large number of missing/null values  
- Categories Synonymous to Null/NaN, but not being read as Null/NaN  
  - "?", 'Unknown/Invalid', 'Not Mapped', 'Not Available', 'NULL' are used in multiple categories, but aren't appearing in the non-null count
- Columns with a high cardinality of categories  
  - admission_source, discharge_disposition, and medical_specialty have too many unique values
  - diag_1, diag_2, and diag_3 also have too many unique values 
- Columns lacking information  
  - examide and citoglipton contain only one unique value  
- Target Variable is split into 3 categories  
  - Readmitted has 3 categories: 'NO', '>30', '<30'    
 
Solutions to the aforementioned problems:
- Drop weight column since most of the weights aren't recorded. A1Cresult and max_glu_serum are important so we might keep them, but will try to work around by redfining "None" values as "No Test Taken" to signify that the values here aren't missing, rather, the patient just didn't take the tests.
- Rename the Null-Synonymous values in all columns to NaN to standardize the value for null in the dataset
- Consolidate the admission_source, discharge_disposition, medical specialty, diag_1, diag_2, and diag_3 columns
- Drop the examide and citoglipton columns
- Leave the readmitted category unchanged for now

#### Merge the IDF Mapping Datasets with the Diabetes Dataset

In [45]:
# we store the merged Diabetes dataset and the IDF mappings into a new dataframe variable in case we need to reference the original diabetes dataset 
# we merge in admissionSource
diabetes_with_idf = pd.merge(DiabetesDF, admissionSourceDF, on = 'admission_source_id')

# we merge in admission type
diabetes_with_idf = pd.merge(diabetes_with_idf, admissionTypeDF, on = 'admission_type_id')

# we merge in discharge disposition
diabetes_with_idf = pd.merge(diabetes_with_idf, dischargeDispositionDF, on = 'discharge_disposition_id')

# rename the the added columns correctly
diabetes_with_idf.rename(columns={'description_x': 'admission_source', 
                                  'description_y': 'admission_type', 'description': 'discharge_disposition'}, inplace=True)

# drop the id columns for each since we don't need it for EDA
diabetes_with_idf.drop(columns=['admission_type_id', 'discharge_disposition_id', 'admission_source_id'], inplace=True)

print(diabetes_with_idf.head(5))

   encounter_id  patient_nbr             race  gender      age weight  \
0       2278392      8222157        Caucasian  Female   [0-10)      ?   
1        149190     55629189        Caucasian  Female  [10-20)      ?   
2         64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3        500364     82442376        Caucasian    Male  [30-40)      ?   
4         16680     42519267        Caucasian    Male  [40-50)      ?   

   time_in_hospital payer_code         medical_specialty  num_lab_procedures  \
0                 1          ?  Pediatrics-Endocrinology                  41   
1                 3          ?                         ?                  59   
2                 2          ?                         ?                  11   
3                 2          ?                         ?                  44   
4                 1          ?                         ?                  51   

   num_procedures  num_medications  number_outpatient  number_emergency  \
0    

#### Drop Columns

We drop the weight column because almost 97% of values are blank.  
We drop the Examide and Citoglipton categories because all patients in the dataset weren't prescribed Examide and Citoglipton so there is no variability in the data 

In [48]:
# Drop 'weight' column from the dataset with idf, since it doesn't give much information 
# axis = 1 basically means drop the entire column. axis = 0 means drop the rows
diabetes_with_idf.drop('weight', axis=1, inplace=True)

# Drop the Examide and Citoglipton categories because none of the patients were prescribed it
diabetes_with_idf.drop('examide', axis=1, inplace=True)
diabetes_with_idf.drop('citoglipton', axis=1, inplace=True) 

#### Replacing "None" for "No Test Taken"

We replace "None" for "No Test Taken" since these values aren't technically null. They mean that the patient wasn't given the test.

In [51]:
diabetes_with_idf['A1Cresult'] = diabetes_with_idf['A1Cresult'].replace([None, 'None'], 'No Test Taken')
diabetes_with_idf['max_glu_serum'] = diabetes_with_idf['max_glu_serum'].replace([None, 'None'], 'No Test Taken')

#### Replacing Null-Synonymous values with NaN

We replace the following values in the dataset with NaN to standardize Null values:
- '?'
- 'Unknown/Invalid'
- 'Not Mapped'
- ' Not Mapped'
- 'Not Available'
- ' Not Available'
- 'NULL'

In [54]:
diabetes_with_idf = diabetes_with_idf.replace(['?', 'Unknown/Invalid', 'Not Mapped', ' Not Mapped', 'NULL', 'Not Available', ' Not Available'], np.nan)

#### Helper Function for Consolidations

In [56]:
# helper function to do the consolidations
def map_groups(colVal):
    # if the value isn't NaN
    if not pd.isna(colVal):
        # we make sure to take into account the white space in front and trailing the string/column value
        colVal = colVal.strip()
        
        # for each grouping, we check if the passed in colVal is in the list of options for each grouping
        # if so, it will return the first match 
        for grouping, options in groupings.items():
            if colVal in options:
                return grouping
    return colVal

#### Consolidating the medical_specialty column

The medical_specialty column contains 84 distinct values. It has been consolidated to 17 categories in which related categories were grouped together. The columns are as follows:
- 'Pediatrics' 
- 'Primary Care/Internal Medicine'
- 'Cardiology/Cardiovascular' 
- 'Surgery' 
- 'Orthopedics/Musculoskeletal'
- 'Gastroenterology' 
- 'Nephrology' 
- 'Behavioral/Mental Health'
- 'Emergency/Trauma' 
- 'Obstetrics/Gynecology' 
- 'Oncology/Hematology'
- 'Other/Rare' 
- 'Endocrinology/Metabolism' 
- 'Neurology/Neurophysiology'
- 'Diagnostics/Imaging/Lab' 
- 'Infectious Disease/Immunology'

In [59]:
# This is the dictionary that contains the new groups : old options  
groupings = {
    # Pediatrics related categories
    'Pediatrics': [ 'Pediatrics', 'Pediatrics-CriticalCare', 'Pediatrics-Pulmonology', 'Pediatrics-Hematology-Oncology', 'Pediatrics-Neurology',
        'Pediatrics-EmergencyMedicine', 'Pediatrics-InfectiousDiseases', 'Pediatrics-AllergyandImmunology', 'Pediatrics-Endocrinology',
        'Anesthesiology-Pediatric', 'Psychiatry-Child/Adolescent', 'Cardiology-Pediatric'],

    # Endocrinology related categories
    'Endocrinology/Metabolism': ['Endocrinology', 'Endocrinology-Metabolism'],

    # Primary Care / Internal Medicine related categories
    'Primary Care/Internal Medicine': [ 'InternalMedicine', 'Family/GeneralPractice', 'Hospitalist', 
                                          'Osteopath', 'Resident', 'PhysicianNotFound', 'Pulmonology'],

    # Cardiology specialties related categories
    'Cardiology/Cardiovascular': ['Cardiology'],

    # Neurology/Neurophysiology related categories
    'Neurology/Neurophysiology': ['Neurology', 'Neurophysiology'],

    # Nephrology related categories
    'Nephrology': ['Nephrology'],

    # All surgery related categories
    'Surgery': ['Surgery-General', 'Surgery-Neuro', 'Surgery-Colon&Rectal', 'Surgery-Plastic', 'Surgery-Thoracic', 
                'Surgery-PlasticwithinHeadandNeck', 'Surgery-Pediatric', 'Surgery-Maxillofacial', 'Surgery-Vascular', 'Surgeon', 
                'SurgicalSpecialty', 'Surgery-Cardiovascular', 'Surgery-Cardiovascular/Thoracic', 'Urology'],

    # Infectious Disease and Immunology related categories
    'Infectious Disease/Immunology': ['InfectiousDiseases', 'AllergyandImmunology'],

    # Obstetrics/Gynecology related categories
    'Obstetrics/Gynecology': ['ObstetricsandGynecology', 'Gynecology', 'Obstetrics', 'Obsterics&Gynecology-GynecologicOnco', 'Perinatology'],

    # Oncology/Hematology related categories
    'Oncology/Hematology': ['Hematology/Oncology', 'Hematology', 'Oncology'],

    # Behavioral/Mental Health related categories
    'Behavioral/Mental Health': ['Psychiatry', 'Psychiatry-Addictive', 'Psychology'],

    # Orthopedics/Musculoskeletal related categories
    'Orthopedics/Musculoskeletal': ['Orthopedics', 'Orthopedics-Reconstructive', 'SportsMedicine', 'PhysicalMedicineandRehabilitation', 'Podiatry'],

    # Diagnostics/Imaging/Lab related categories
    'Diagnostics/Imaging/Lab': ['Radiology', 'Radiologist', 'Pathology'],
	
	# Gastroenterology related categories
	'Gastroenterology': ['Gastroenterology'],
	
	# Emergency/Trauma related categories
	'Emergency/Trauma': ['Emergency/Trauma'],

    # All other other/rare specialties 
    'Other/Rare': ['Otolaryngology', 'Speech', 'OutreachServices', 'DCPTEAM', 
                     'Proctology', 'Rheumatology', 'Dentistry', 'Dermatology', 'Anesthesiology', 'Ophthalmology']
}

# we apply the mapping and store this consolidation in a new column
diabetes_with_idf['medical_specialty_consol'] = diabetes_with_idf['medical_specialty'].apply(map_groups)

# just checking if it worked
print(diabetes_with_idf['medical_specialty_consol'].unique())
print(len(diabetes_with_idf['medical_specialty_consol'].unique()))

['Pediatrics' nan 'Primary Care/Internal Medicine'
 'Cardiology/Cardiovascular' 'Surgery' 'Orthopedics/Musculoskeletal'
 'Gastroenterology' 'Nephrology' 'Behavioral/Mental Health'
 'Emergency/Trauma' 'Obstetrics/Gynecology' 'Oncology/Hematology'
 'Other/Rare' 'Endocrinology/Metabolism' 'Neurology/Neurophysiology'
 'Diagnostics/Imaging/Lab' 'Infectious Disease/Immunology']
17


#### Consolidating diag_1, diag_2, diag_3 columns

Instead of having >800 codes to diagnose the patient being admitted, the codes are consolidated into <50 generalized categories based on their codes. 
- Numeric Codes represent: Diseases and Disorders diagnosed to patient
- V Codes represent: Reasons for a patient to come in (preventative care, routine check-ups, etc.)
- E Codes represent: External causes of Injury

All Numeric Codes will be mapped to their corresponding codes according to this source:  
- https://www.aapc.com/codes/icd9-codes-range/
  
All V Codes will be mapped to their corresponding codes according to this source:  
- https://www.aapc.com/codes/icd9-codes-range/136/
  
All E Codes will be just labeled as "E-Codes: External Causes of Injury and Poison" since they're external factors that don't usually factor in readmissions in diabetes patients

##### Helper function convert ICD-9 Diagnostic codes to the category we want

In [63]:
def icd9_to_category(code):
    # case if we find a NaN entry
    if pd.isna(code):
        return code

    # we first try to catch the codes with a V or E
    if code.startswith('V'):
        # we get the number part of the code
        code = code[1:]
        # convert it to an int for comparison purposes
        code = int(float(code))

        # converting the code to its corresponding v code
        if 1 <= code <= 6:
            return "V-Code: Communicable Disease Hazards"
        elif 7 <= code <= 9:
            return "V-Code: Isolation/Prophylactic Measures/Other Hazards"
        elif 10 <= code <= 19:
            return "V-Code: Personal/Family Health Hazards"
        elif 20 <= code <= 29:
            return "V-Code: Reproduction & Development"
        elif 30 <= code <= 39:
            return "V-Code: Type of Birth"
        elif 40 <= code <= 49:
            return "V-Code: Condition Influencing Health Status"
        elif 50 <= code <= 59:
            return "V-Code: Specific Procedures & Aftercare"
        elif 60 <= code <= 69:
            return "V-Code: Other Health Service Circumstances"
        elif 70 <= code <= 82:
            return "V-Code: Examination & Investigation (No Diagnosis)"
        elif 83 <= code <= 84:
            return "V-Code: Genetics"
        elif code == 85:
            return "V-Code: Body Mass Index"
        elif code == 86:
            return "V-Code: Estrogen Receptor Status"
        elif code == 87:
            return "V-Code: Other Personal Exposures"
        elif code == 88:
            return "V-Code: Acquired Absence of Organs/Tissue"
        elif code == 89:
            return "V-Code: Suspected Conditions Not Found"
        elif code == 90:
            return "V-Code: Retained Foreign Body"
        elif code == 91:
            return "V-Code: Multiple Gestation Placenta Status"
        else:
            return 'Other V-Code'
    elif code.startswith('E'):
        # since E codes are external causes and don't really play a role in readmissions, we won't 
        return 'E-Codes: External Causes of Injury and Poison'

    # we convert the code from a string/object to an num for comparison 
    # we don't need to get into specifics so we get the int
    code = int(float(code))

    # This performs the mapping for the codes that don't start with E or V
    if 1 <= code <= 139:
        return 'INFECTIOUS/PARASITIC'
    elif 140 <= code <= 239:
        return 'NEOPLASMS'
    elif 240 <= code <= 248:
        return 'ENDOCRINE/NUTRITIONAL/METABOLIC/IMMUNITY'
    elif 249 <= code <= 250:
        return 'Diabetes Mellitus'
    elif 251 <= code <= 279:
        return 'ENDOCRINE/NUTRITIONAL/METABOLIC/IMMUNITY'
    elif 280 <= code <= 289:
        return 'BLOOD/BLOOD-FORMING ORGANS'
    elif 290 <= code <= 319:
        return 'MENTAL/BEHAVIORAL/NEURODEVELOPMENTAL'
    elif 320 <= code <= 389:
        return 'NERVOUS SYSTEM/SENSE ORGANS'
    elif 390 <= code <= 459:
        return 'CIRCULATORY SYSTEM'
    elif 460 <= code <= 519:
        return 'RESPIRATORY SYSTEM'
    elif 520 <= code <= 579:
        return 'DIGESTIVE SYSTEM'
    elif 580 <= code <= 629:
        return 'GENITOURINARY SYSTEM'
    elif 630 <= code <= 679:
        return 'PREGNANCY/CHILDBIRTH/PUERPERIUM'
    elif 680 <= code <= 709:
        return 'SKIN/SUBCUTANEOUS TISSUE'
    elif 710 <= code <= 739:
        return 'MUSCULOSKELETAL SYSTEM/CONNECTIVE TISSUE'
    elif 740 <= code <= 759:
        return 'CONGENITAL ANOMALIES'
    elif 760 <= code <= 779:
        return 'PERINATAL PERIOD CONDITIONS'
    elif 780 <= code <= 799:
        return 'SYMPTOMS/SIGNS/ILL-DEFINED CONDITIONS'
    elif 800 <= code <= 999:
        return 'INJURY/POISONING'
    else:
        return 'Unknown Code'

##### Applying the Consolidation

In [65]:
# we apply the mapping to all 3 diagnostic columns and store each consolidation in a new column
diabetes_with_idf['diag_1_consol'] = diabetes_with_idf['diag_1'].apply(icd9_to_category)
diabetes_with_idf['diag_2_consol'] = diabetes_with_idf['diag_2'].apply(icd9_to_category)
diabetes_with_idf['diag_3_consol'] = diabetes_with_idf['diag_3'].apply(icd9_to_category)

# just checking if it worked
print('diag_1_consol')
print(diabetes_with_idf['diag_1_consol'].unique())
print(len(diabetes_with_idf['diag_1_consol'].unique()))
print('\n')

print('diag_2_consol')
print(diabetes_with_idf['diag_2_consol'].unique())
print(len(diabetes_with_idf['diag_2_consol'].unique()))
print('\n')

print('diag_3_consol')
print(diabetes_with_idf['diag_3_consol'].unique())
print(len(diabetes_with_idf['diag_3_consol'].unique()))
print('\n')

diag_1_consol
['Diabetes Mellitus' 'ENDOCRINE/NUTRITIONAL/METABOLIC/IMMUNITY'
 'PREGNANCY/CHILDBIRTH/PUERPERIUM' 'INFECTIOUS/PARASITIC' 'NEOPLASMS'
 'CIRCULATORY SYSTEM' 'RESPIRATORY SYSTEM' 'INJURY/POISONING'
 'SKIN/SUBCUTANEOUS TISSUE' 'MUSCULOSKELETAL SYSTEM/CONNECTIVE TISSUE'
 'DIGESTIVE SYSTEM' 'V-Code: Specific Procedures & Aftercare'
 'SYMPTOMS/SIGNS/ILL-DEFINED CONDITIONS' 'GENITOURINARY SYSTEM'
 'MENTAL/BEHAVIORAL/NEURODEVELOPMENTAL' 'NERVOUS SYSTEM/SENSE ORGANS'
 'BLOOD/BLOOD-FORMING ORGANS' nan 'CONGENITAL ANOMALIES'
 'V-Code: Condition Influencing Health Status'
 'V-Code: Other Health Service Circumstances'
 'V-Code: Reproduction & Development'
 'V-Code: Examination & Investigation (No Diagnosis)'
 'E-Codes: External Causes of Injury and Poison'
 'V-Code: Isolation/Prophylactic Measures/Other Hazards']
25


diag_2_consol
[nan 'Diabetes Mellitus' 'NEOPLASMS' 'CIRCULATORY SYSTEM'
 'RESPIRATORY SYSTEM' 'BLOOD/BLOOD-FORMING ORGANS' 'INJURY/POISONING'
 'MUSCULOSKELETAL SYSTEM/CO

#### Consolidating the admission_source column

The admission_source column contains 21 distinct values. It has been consolidated to 17 categories in which related categories were grouped together. The new categories are as follows:
- Referral/Planned Admission
- Transfer from Short-Term Facil.
- Transfer from Long-Term  Facil.
- Emergency/Unplanned Admission
- Birth-related Admissions

In [68]:
# This is the dictionary that contains the new groups : old options  
groupings = {
    "Referral/Planned Admission": [
        "Physician Referral",
        "Clinic Referral",
        "HMO Referral"
    ],
    "Transfer from Short-Term Facil.": [
        "Transfer from a hospital",
        "Transfer from critial access hospital",
        "Transfer from another health care facility",
        "Transfer from hospital inpt/same fac reslt in a sep claim",
        "Transfer from Ambulatory Surgery Center"
    ],
    "Transfer from Long-Term  Facil.": [
        "Transfer from a Skilled Nursing Facility (SNF)",
        "Transfer From Another Home Health Agency",
        "Readmission to Same Home Health Agency",
        "Transfer from Hospice"
    ],
    "Emergency/Unplanned Admission": [
        "Emergency Room",
        "Court/Law Enforcement"
    ],
    "Birth-related Admissions": [
        "Normal Delivery",
        "Premature Delivery",
        "Sick Baby",
        "Extramural Birth",
        "Born inside this hospital",
        "Born outside this hospital"
    ],
}

# we apply the mapping and store this consolidation in a new column
diabetes_with_idf['admission_source_consol'] = diabetes_with_idf['admission_source'].apply(map_groups)

# just checking if it worked
print(diabetes_with_idf['admission_source_consol'].unique())
print(len(diabetes_with_idf['admission_source_consol'].unique()))

['Referral/Planned Admission' 'Emergency/Unplanned Admission'
 'Transfer from Short-Term Facil.' 'Transfer from Long-Term  Facil.' nan
 'Birth-related Admissions']
6


#### Consolidating the discharge_disposition column

The discharge_disposition column contains 29 distinct values. It has been consolidated to 8 categories in which related categories were grouped together. The categories are as follows:
- Home
- Outpatient Care
- Short-term / Inpatient Care
- Long-term Care and Nursing Facilities
- Hospice / Death
- AMA
- Other / Specialty

In [71]:
# This is the dictionary that contains the new groups : old options  
groupings = {
    "Home": [
        "Discharged to home",
        "Discharged/transferred to home with home health service",
        "Discharged/transferred to home under care of Home IV provider"
    ],
    "Outpatient Care": [
        "Still patient or expected to return for outpatient services",
        "Discharged/transferred/referred another institution for outpatient services",
        "Discharged/transferred/referred to this institution for outpatient services",
        "Neonate discharged to another hospital for neonatal aftercare"
    ],
    "Short-term / Inpatient Care": [
        "Discharged/transferred to another short term hospital",
        "Discharged/transferred to another type of inpatient care institution",
        "Discharged/transferred within this institution to Medicare approved swing bed",
        "Discharged/transferred to a Critical Access Hospital (CAH).",
        "Discharged/transferred to another rehab fac including rehab units of a hospital .",
        "Admitted as an inpatient to this hospital"
    ],
    "Long-term Care and Nursing Facilities": [
        "Discharged/transferred to SNF",
        "Discharged/transferred to ICF",
        "Discharged/transferred to a long term care hospital.",
        "Discharged/transferred to a nursing facility certified under Medicaid but not certified under Medicare."
    ],
    "Hospice / Death": [
        "Expired",
        "Hospice / home",
        "Hospice / medical facility",
        "Expired at home. Medicaid only, hospice.",
        "Expired in a medical facility. Medicaid only, hospice.",
        "Expired, place unknown. Medicaid only, hospice."
    ],
    "AMA": [
        "Left AMA"
    ],
    "Other / Specialty": [
        "Discharged/transferred to another Type of Health Care Institution not Defined Elsewhere",
        "Discharged/transferred to a federal health care facility.",
        "Discharged/transferred/referred to a psychiatric hospital of psychiatric distinct part unit of a hospital"
    ]
}

# we apply the mapping and store this consolidation in a new column
diabetes_with_idf['discharge_disposition_consol'] = diabetes_with_idf['discharge_disposition'].apply(map_groups)

# just checking if it worked
print(diabetes_with_idf['discharge_disposition_consol'].unique())
print(len(diabetes_with_idf['discharge_disposition_consol'].unique()))

[nan 'Home' 'Long-term Care and Nursing Facilities'
 'Short-term / Inpatient Care' 'Hospice / Death' 'AMA' 'Outpatient Care'
 'Other / Specialty']
8


#### Looking at the Cleaned Dataframe for EDA

In [73]:
print(diabetes_with_idf.info())
print(diabetes_with_idf.head(5))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 53 columns):
 #   Column                        Non-Null Count   Dtype 
---  ------                        --------------   ----- 
 0   encounter_id                  101766 non-null  int64 
 1   patient_nbr                   101766 non-null  int64 
 2   race                          99493 non-null   object
 3   gender                        101763 non-null  object
 4   age                           101766 non-null  object
 5   time_in_hospital              101766 non-null  int64 
 6   payer_code                    61510 non-null   object
 7   medical_specialty             51817 non-null   object
 8   num_lab_procedures            101766 non-null  int64 
 9   num_procedures                101766 non-null  int64 
 10  num_medications               101766 non-null  int64 
 11  number_outpatient             101766 non-null  int64 
 12  number_emergency              101766 non-null  int64 
 13 

#### For Output Text Information File

In [75]:
# Print drop reason.
removing_weight_column += "Dropped 'weight' column due to too many missing elements (97% missing).\n\n"

## EDA

## Modeling & Analysis

### Encoding - Preparing the dataset for the model

In [79]:
# We also drop 'weight' column for the original dataset, since it doesn't give much information 
DiabetesDF.drop('weight', axis=1, inplace=True)

In [80]:
# Replace race text with integers.
replacing_race_with_integers += "Column Name: Race\n"
# Print replacement information.
replacing_race_with_integers += "Replacements:\n'?' --> 0\n'Caucasian' --> 1\n'AfricanAmerican' --> 2\n'Asian' --> 3\n'Hispanic' --> 4\n'Other' --> 5\n'Anything Else' --> -1\n"
# Print list of unique elements beforehand.
replacing_race_with_integers += unique_element_and_count_string("race", "before")
# Replacement function.
def replace_race_text(input_information):
    match input_information:
        case '?': # If '?' (not provided), return 0.
            return 0
        case 'Caucasian':
            return 1
        case 'AfricanAmerican':
            return 2
        case 'Asian':
            return 3
        case 'Hispanic':
            return 4
        case 'Other':
            return 5
        case _:
            return -1
# Perform replacement.
DiabetesDF['race'] = DiabetesDF['race'].apply(lambda x: replace_race_text(x))
# Print list of unique elements afterward to ensure replacement.
replacing_race_with_integers += unique_element_and_count_string("race", "after")
replacing_race_with_integers += "\n"

In [81]:
# Replace gender text with integers.
replacing_gender_with_integers += "Column Name: gender\n"
# Print replacement information.
replacing_gender_with_integers += "Replacements:\n'Unknown/Invalid' --> 0\n'Male' --> 1\n'Female' --> 2\n'Anything Else' --> -1\n"
# Print list of unique elements beforehand.
replacing_gender_with_integers += unique_element_and_count_string("gender", "before")
# Replacement function.
def replace_gender(input_information):
    match input_information:
        case 'Unknown/Invalid':
            return 0
        case 'Male':
            return 1
        case 'Female':
            return 2
        case _:
            return -1
# Perform replacement.
DiabetesDF['gender'] = DiabetesDF['gender'].apply(lambda x: replace_gender(x))
# Print list of unique elements afterward to ensure replacement.
replacing_gender_with_integers += unique_element_and_count_string("gender", "after")
replacing_gender_with_integers += "\n"

In [82]:
# Replace age text with integers.
replace_age_categories_with_integers += "Column Name: age\n"
# Print replacement information.
replace_age_categories_with_integers += "Replacements:\n'[0-10)' --> 0\n'[10-20)' --> 1\n'[20-30)' --> 2\n'[30-40)' --> 3\n'[40-50)' --> 4\n'[50-60)' --> 5\n'[60-70)' --> 6\n'[70-80)' --> 7\n'[80-90)' --> 8\n'[90-100)' --> 9\n'Anything Else' --> -1\n"
# Print list of unique elements beforehand.
replace_age_categories_with_integers += unique_element_and_count_string("age", "before")
# Replacement function.
def replace_age_ranges(input_information):
    match input_information:
        case '[0-10)':
            return 0
        case '[10-20)':
            return 1
        case '[20-30)':
            return 2
        case '[30-40)':
            return 3
        case '[40-50)':
            return 4
        case '[50-60)':
            return 5
        case '[60-70)':
            return 6
        case '[70-80)':
            return 7
        case '[80-90)':
            return 8
        case '[90-100)':
            return 9
        case _:
            return -1
# Perform replacement.
DiabetesDF['age'] = DiabetesDF['age'].apply(lambda x: replace_age_ranges(x))
# Print list of unique elements afterward to ensure replacement.
replace_age_categories_with_integers += unique_element_and_count_string("age", "after")
replace_age_categories_with_integers += "\n"

In [83]:
# Condense insurance information.
condensed_insurance_information += "Column Name: payer_code\n"
# Print condensing information.
condensed_insurance_information += "Condensing:\n'?' --> 0\n'SP' --> 1\n'Anything Else' --> 2\n"
# Print reason for condensing.
condensed_insurance_information += "Reason for condensing:\nPatients who don't use insurance are less likely to return for additional treatment, which puts them at increased risk for further injury.\n"
# Print list of unique elements beforehand.
condensed_insurance_information += unique_element_and_count_string("payer_code", "before")
# Filtering function.
def filter_payer_code(input_information):
    # Insurance information not provided.
    if (input_information == '?'):
        return 0
    # Patient did not use insurance. I don't know if 'SP' stands for 'self-pay' or not.
    elif (input_information == 'SP'):
        return 1
    # Patient used insurance.
    else:
        return 2
# Perform the filtering operation.
DiabetesDF['payer_code'] = DiabetesDF['payer_code'].apply(lambda x: filter_payer_code(x))
# Print list of unique elements afterward to ensure filtering.
condensed_insurance_information += unique_element_and_count_string("payer_code", "after")
condensed_insurance_information += "\n"

In [84]:
# Replace the elements in a given column with integers.
def replace_elements_with_ints(column_name):
    return_question_mark_associated_element = 0

    # Get the list of unique column elements from the column in question.
    unique_column_elements = DiabetesDF[column_name].unique().tolist()

    # Check if a '?' element exists, and move it to the front of the list of unique column elements if it does.
    try:
        question_mark_location = unique_column_elements.index('?')
        unique_column_elements.pop(question_mark_location)
        unique_column_elements.insert(0, '?')

        # Additional information needs to be printed
        if (question_mark_location == 0):
            return_question_mark_associated_element = 1
        if (question_mark_location > 0):
            return_question_mark_associated_element = 2
    except:
        pass

    # Convert back to numpy array for printing purposes.
    unique_column_elements = np.array(unique_column_elements)

    # Get the number of unique column elements from the column in question, and then generate a list of elements from 0 up to one less than the number of unique elements.
    unique_column_element_count = DiabetesDF[column_name].nunique()
    list_of_numbers = list(range(0, unique_column_element_count))

    # Pair each number to a unique column element and perform the replacement.
    unique_column_elements_numbers_pairing = {key: element for key, element in zip(unique_column_elements, list_of_numbers)}
    DiabetesDF[column_name] = DiabetesDF[column_name].apply(lambda x: lambda x: unique_column_elements_numbers_pairing[x])

    # If a '?' element exists, the temporary reordering of unique elements was done, so additional information needs to be printed.
    if (return_question_mark_associated_element == 1):
        return f"The element associated with '?' is '{unique_column_elements_numbers_pairing['?']}'\n"
    elif (return_question_mark_associated_element == 2):
        return (
            f"Temporary reordering of unique elements: {unique_column_elements}\n" +
            f"The element associated with '?' is '{unique_column_elements_numbers_pairing['?']}'\n"
            )
    else:
        return ""
# Replacement information.
replace_elements_with_ints_string = "Replacement:\nEverything --> After moving '?' (if it exists) to the front of the unique elements list, each element is assigned a unique number, starting from zero up to one less than the number of unique elements.\n"


In [85]:
# Replacing medical specialty strings with integers.
# Despite 53% of element missing, the medical speciality examining a patient is very important.
#  Example: Emergency/Trauma is much more important than Dentistry.
def consensing_medical_specialties(input_information):
    match input_information:
        # Primary Care
        case x if (x == 'Family/GeneralPractice') or (x == 'InternalMedicine') or (x == 'Pediatrics') or (x == 'Hospitalist') or (x == 'Osteopath') or (x == 'Resident') or (x == 'PhysicianNotFound'):
            return 'Primary Care'
        # Cardiovascular System - Non-surgery.
        case x if (x == 'Cardiology') or (x == 'Cardiology-Pediatric'):
            return 'Cardiovascular System - Non-surgery'
        # Neurology - Non-surgery.
        case x if (x == 'Neurology') or (x == 'Neurophysiology') or (x == 'Pediatrics-Neurology'):
            return 'Neurology - Non-surgery'
        # Psychiatry
        case x if (x == 'Psychiatry') or (x == 'Psychiatry-Child/Adolescent') or (x == 'Psychiatry-Addictive') or (x == 'Psychology'):
            return 'Psychiatry'
        case _:
            return input_information
# Condense some of the related medical specialities into groups.
replacing_medical_specialty_with_integers += "Column Name: medical_specialty\n"
# Perform consensing.
DiabetesDF['medical_specialty'] = DiabetesDF['medical_specialty'].apply(lambda x: consensing_medical_specialties(x))
# Print list and number of unique elements after condensing.
replacing_medical_specialty_with_integers += unique_element_and_count_string('medical_specialty', "After Condensing")
# Print notes.
replacing_medical_specialty_with_integers += ("Notes:\nDespite 53% of elements missing, the medical speciality examining a patient is very important.\n" +
                                              "Example:\nEmergency/Trauma is much more important than Dentistry.\n" + 
                                              "In addition, 'DCPTEAM' possibly stands for 'Dynamic Care Planning Team'.\n")
# Print replacement message.
replacing_medical_specialty_with_integers += replace_elements_with_ints_string
# Print list and number of unique elements beforehand.
replacing_medical_specialty_with_integers += unique_element_and_count_string("medical_specialty", "before")
# Perform replacement.
replacing_medical_specialty_with_integers += replace_elements_with_ints('medical_specialty')
# Print list and number of unique elements afterward to ensure replacement.
replacing_medical_specialty_with_integers += unique_element_and_count_string("medical_specialty", "after")
replacing_medical_specialty_with_integers += "\n"

In [86]:
# Replace the codes for diagnoses 1, 2, and 3 with integers.
for i in range(18, 21):
    current_column_name = column_names_list[i]
    # Print column name.
    replacing_diagnoses_codes_with_integers += f"Column Name: {current_column_name}\n"
    # Print reason for replacement.
    replacing_diagnoses_codes_with_integers += replace_elements_with_ints_string
    # Print list and number of unique elements beforehand.
    replacing_diagnoses_codes_with_integers += unique_element_and_count_string(current_column_name, "before")
    # Perform replacement.
    replacing_diagnoses_codes_with_integers += replace_elements_with_ints(current_column_name)
    # Print list and number of unique elements afterward to ensure replacement.
    replacing_diagnoses_codes_with_integers += unique_element_and_count_string(current_column_name, "after")
    replacing_diagnoses_codes_with_integers += "\n"


In [87]:
# Replace no, steady, up, and down with corresponding integers.
def replace_no_steady_up_down(input_information):
    match input_information.lower():
        case 'no':
            return 0
        case 'steady':
            return 1
        case 'up':
            return 2
        case 'down':
            return 3
        case _:
            return -1
# Print medication note.
replacing_medication_used_indication_with_integers += "Medication Note:\nWhile some medications were not administered, medications are still important and, as a result, they weren't dropped.\n\n"
for i in range(24, 47):
    current_column_name = column_names_list[i]
    # Print column name.
    replacing_medication_used_indication_with_integers += f"Column Name: {current_column_name}\n"
    # Print reason for replacement.
    replacing_medication_used_indication_with_integers += "Replacements:\n'no' --> 0\n'steady' --> 1\n'up' --> 2\n'down' --> 3\n'Anything Else' --> -1\n"
    # Print list and number of unique elements beforehand.
    replacing_medication_used_indication_with_integers += unique_element_and_count_string(current_column_name, "before")
    # Perform replacement.
    DiabetesDF[current_column_name] = DiabetesDF[current_column_name].apply(lambda x: replace_no_steady_up_down(x))
    # Print list and number of unique elements afterward to ensure replacement.
    replacing_medication_used_indication_with_integers += unique_element_and_count_string(current_column_name, "after")
    replacing_medication_used_indication_with_integers += "\n"

In [88]:
# Replace elements in 'max_glu_serum' column.
replacing_max_glu_serum_with_integers += "Column Name: max_glu_serum\n"
# Print replacement information.
replacing_max_glu_serum_with_integers += "Replacements:\n'nan' --> 0\n'norm' --> 1\n'>200' --> 2\n'>300' --> 3\n'Anything Else' --> -1\n"
# Print note.
replacing_max_glu_serum_with_integers += "Note:\n'None' in the excel sheet might be 'nan' in Pandas. That explains the jump from 3 to 4 in the number of unique elements.\n"
# Print list and number of unique elements beforehand.
replacing_max_glu_serum_with_integers += unique_element_and_count_string("max_glu_serum", "before")
# Replace elements with corresponding integers.
def replace_max_glu_serum_elements(input_information):
    match str(input_information).lower():
        case 'nan':
            return 0
        case 'norm':
            return 1
        case '>200':
            return 2
        case '>300':
            return 3
        case _:
            return -1
# Perform replacement.
DiabetesDF['max_glu_serum'] = DiabetesDF['max_glu_serum'].apply(lambda x: replace_max_glu_serum_elements(x))
# Print list and number of unique elements afterward to ensure replacement.
replacing_max_glu_serum_with_integers += unique_element_and_count_string("max_glu_serum", "after")
replacing_max_glu_serum_with_integers += "\n"

In [89]:
# Replace elements in 'A1Cresult' column.
replacing_A1Cresult_with_integers += "Column Name: A1Cresult\n"
# Perform replacement information.
replacing_A1Cresult_with_integers += "Replacements:\n'nan' --> 0\n'norm' --> 1\n'>7' --> 2\n'>8' --> 3\n'Anything Else' --> -1\n"
# Print note.
replacing_A1Cresult_with_integers += "Note:\n'None' in the excel sheet might be 'nan' in Pandas. That explains the jump from 3 to 4 in the number of unique elements.\n"
# Print list and number of unique elements beforehand.
replacing_A1Cresult_with_integers += unique_element_and_count_string("A1Cresult", "before")
# Replace elements with corresponding integers.
def replace_A1Cresult_elements(input_information):
    match str(input_information).lower():
        case 'nan':
            return 0
        case 'norm':
            return 1
        case '>7':
            return 2
        case '>8':
            return 3
        case _:
            return -1
# Perform replacement.
DiabetesDF['A1Cresult'] = DiabetesDF['A1Cresult'].apply(lambda x: replace_A1Cresult_elements(x))
# Print list and number of unique elements afterward to ensure replacement.
replacing_A1Cresult_with_integers += unique_element_and_count_string("A1Cresult", "after")
replacing_A1Cresult_with_integers += "\n"

In [90]:
# Replace elements in 'change' column.
replacing_change_with_integers += "Column Name: change\n"
# Print replacement information.
replacing_change_with_integers += "Replacements:\n'no' --> 0\n'ch' --> 1\n'Anything Else' --> -1\n"
# Print list and number of unique elements beforehand.
replacing_change_with_integers += unique_element_and_count_string("change", "before")
# Replace elements with corresponding integers.
def replace_change_elements(input_information):
    match str(input_information).lower():
        case 'no':
            return 0
        case 'ch':
            return 1
        case _:
            return -1
# Perform replacement.
DiabetesDF['change'] = DiabetesDF['change'].apply(lambda x: replace_change_elements(x))
# Print list and number of unique elements afterward to ensure replacement.
replacing_change_with_integers += unique_element_and_count_string("change", "after")
replacing_change_with_integers += "\n"


In [91]:
# Replace elements in 'diabetesMed' column.
replacing_diabetesMed_with_integers += "Column Name: diabetesMed\n"
# Print replacement information.
replacing_diabetesMed_with_integers += "Replacements:\n'no' --> 0\n'yes' --> 1\n'Anything Else' --> -1\n"
# Print list and number of unique elements beforehand.
replacing_diabetesMed_with_integers += unique_element_and_count_string("diabetesMed", "before")
# Replace elements with corresponding integers.
def replace_diabetesMed_elements(input_information):
    match str(input_information).lower():
        case 'no':
            return 0
        case 'yes':
            return 1
        case _:
            return -1
# Perform replacement.
DiabetesDF['diabetesMed'] = DiabetesDF['diabetesMed'].apply(lambda x: replace_diabetesMed_elements(x))
# Print list and number of unique elements afterward to ensure replacement.
replacing_diabetesMed_with_integers += unique_element_and_count_string("diabetesMed", "after")
replacing_diabetesMed_with_integers += "\n"


In [92]:
# Replace elements in 'readmitted' column.
replacing_readmitted_with_integers += "Column Name: readmitted\n"
# Print replacement information.
replacing_readmitted_with_integers += "Replacements:\n'no' --> 0\n'<30' --> 1\n'>30' --> 2\n'Anything Else' --> -1\n"
# Print list and number of unique elements beforehand.
replacing_readmitted_with_integers += unique_element_and_count_string("readmitted", "before")
# Replace elements with corresponding integers.
def replace_readmitted_elements(input_information):
    match str(input_information).lower():
        case 'no':
            return 0
        case '<30':
            return 1
        case '>30':
            return 2
        case _:
            return -1
# Perform replacement.
DiabetesDF['readmitted'] = DiabetesDF['readmitted'].apply(lambda x: replace_readmitted_elements(x))
# Print list and number of unique elements afterward to ensure replacement.
replacing_readmitted_with_integers += unique_element_and_count_string("readmitted", "after")
replacing_readmitted_with_integers += "\n"


## Creating the Encoded CSV Files and Text Information Files

In [94]:
# print("Outputting information to text files.")
# final_output_string_1 = (
#     "Below, please find a list of the unique elements in each column along with how many unique elements there are.\n\n\n"
#     f"{list_of_unique_and_nunique}\n\n"
# )
# final_output_string_2 = (
#     "Introductory Info:\n"
#     "This file: Information on the modifications done to the diabetic_data.csv file and why they were done.\n"
#     "unique_information.txt: A list of unique elements in each column along with how many unique elements there are.\n"
#     "output.csv: A version of diabetic_data.csv modified with the edits described in this file.\n"
#     "output_grouped.csv: A version of output.csv where everything has been grouped together based on patient_nbr.\n"
#     "Percentages for the amount of missing elements for some of the columns can be found here: https://onlinelibrary.wiley.com/doi/10.1155/2014/781670\n\n\n\n\n"
#     "Modifications to diabetic_data.csv:\n\n\n"
#     f"{check_duplicates_in_first_two}\n\n"
#     f"{replacing_race_with_integers}\n\n"
#     f"{replacing_gender_with_integers}\n\n"
#     f"{replace_age_categories_with_integers}\n\n"
#     f"{removing_weight_column}\n\n"
#     f"{condensed_insurance_information}\n\n"
#     f"{replacing_medical_specialty_with_integers}\n\n"
#     f"{replacing_diagnoses_codes_with_integers}\n\n"
#     f"{replacing_medication_used_indication_with_integers}\n\n"
#     f"{replacing_max_glu_serum_with_integers}\n\n"
#     f"{replacing_A1Cresult_with_integers}\n\n"
#     f"{replacing_change_with_integers}\n\n"
#     f"{replacing_diabetesMed_with_integers}\n\n"
#     f"{replacing_readmitted_with_integers}"
# )

In [95]:
# # Save unique element and count information to text file.
# with open("unique_information.txt", 'w') as file:
#     file.write(final_output_string_1)
# # Save explanation information to text file.
# with open("explanation_information.txt", 'w') as file:
#     file.write(final_output_string_2)
# print("Information outputted to text file.")

In [96]:
# # Create modified csv file from resulting dataframe.
# print("Outputting modified csv file.")
# DiabetesDF.to_csv(os.path.join(file_save_loaction, 'output.csv'), index=False)
# print("Modified csv file outputted.")

In [97]:
# # Group resulting dataframe by the 'patient_nbr' column.
# print("Starting Grouping")
# tqdm.pandas()
# # https://stackoverflow.com/questions/60963409/creating-a-dictionary-of-dictionaries-from-groupby
# DiabetesDF_grouped = DiabetesDF.groupby('patient_nbr')
# group_of_groups = DiabetesDF_grouped.progress_apply(lambda x: x.set_index('encounter_id').to_dict())
# print("Grouping Done")

In [98]:
# # Create grouped csv file from the grouped dataframe.
# print("Outputting grouped version of modified csv file.")
# group_of_groups.to_csv(os.path.join(file_save_loaction, 'output_grouped.csv'), index=False)
# print("Grouped version of modified csv file outputted.")